agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,457 @@
+func (a *Autopilot) MinRaftProtocol() (int, error) {
+	minVersion := -1
+	members := a.delegate.Serf().Members()
+	for _, m := range members {
+		if m.Status != serf.StatusAlive {
+			continue
+		}
+
+		if !a.validServerFunc(m) {
+			continue
+		}
+
+		vsn, ok := m.Tags["raft_vsn"]
+		if !ok {
+			vsn = "1"
+		}
+		raftVsn, err := strconv.Atoi(vsn)
+		if err != nil {
+			return -1, err
+		}
+
+		if minVersion == -1 || raftVsn < minVersion {
+			minVersion = raftVsn
+		}
+	}
+
+	if minVersion == -1 {
+		return minVersion, fmt.Errorf("No servers found")
+	}
+
+	return minVersion, nil
+}
+
+// handlePromotions is a helper shared with Consul Enterprise that attempts to
+// apply desired server promotions to the Raft configuration.
+func (a *Autopilot) handlePromotions(promotions []raft.Server) error {
+	// This used to wait to only promote to maintain an odd quorum of
+	// servers, but this was at odds with the dead server cleanup when doing
+	// rolling updates (add one new server, wait, and then kill an old
+	// server). The dead server cleanup would still count the old server as
+	// a peer, which is conservative and the right thing to do, and this
+	// would wait to promote, so you could get into a stalemate. It is safer
+	// to promote early than remove early, so by promoting as soon as
+	// possible we have chosen that as the solution here.
+	for _, server := range promotions {
+		a.logger.Printf("[INFO] autopilot: Promoting %s to voter", fmtServer(server))
+		addFuture := a.delegate.Raft().AddVoter(server.ID, server.Address, 0, 0)
+		if err := addFuture.Error(); err != nil {
+			return fmt.Errorf("failed to add raft peer: %v", err)
+		}
+	}
+
+	// If we promoted a server, trigger a check to remove dead servers.
+	if len(promotions) > 0 {
+		select {
+		case a.removeDeadCh <- struct{}{}:
+		default:
+		}
+	}
+	return nil
+}
+
+// ServerHealthLoop monitors the health of the servers in the cluster
+func (a *Autopilot) ServerHealthLoop(shutdownCh <-chan struct{}) {
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(a.healthInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-shutdownCh:
+			return
+		case <-ticker.C:
+			if err := a.updateClusterHealth(); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error updating cluster health: %s", err)
+			}
+		}
+	}
+}
+
+// updateClusterHealth fetches the Raft stats of the other servers and updates
+// s.clusterHealth based on the configured Autopilot thresholds
+func (a *Autopilot) updateClusterHealth() error {
+	// Don't do anything if the min Raft version is too low
+	minRaftProtocol, err := a.MinRaftProtocol()
+	if err != nil {
+		return fmt.Errorf("error getting server raft protocol versions: %s", err)
+	}
+	if minRaftProtocol < 3 {
+		return nil
+	}
+
+	autopilotConf, ok := a.delegate.GetOrCreateAutopilotConfig()
+	if !ok {
+		return nil
+	}
+	// Bail early if autopilot config hasn't been initialized yet
+	if autopilotConf == nil {
```

Comment:
If it can return `ok = true` but still have a `nil` config maybe we should just get rid of the Boolean part of the return value and just always check for `nil` to simplify the interface.